### PR TITLE
Add support for dumping Eclipse Locks

### PR DESCRIPTION
**Pull request by briandealwis**
The Eclipse Jobs framework provides means for scheduling work in the form of explicit locks and _scheduling rules_, which it treats as a lock.  The state of these locks are managed by the `IJobManager`.  Unfortunately it doesn't expose any information as to the current state.  But internally it uses a class called `DeadlockDetector` that provides a rather cryptic report via its `toDebugString()`.

This patch to the `ThreadDumpingWatchdog` attempts to obtain this `DeadlockDetector` and output its `toDebugString()`.  The output represents a matrix where the columns are the allocated locks and the rows correspond to the threads and their interest in the locks.
```
| Eclipse Locks:
|  ::
|  R/, OrderedLock (4),
|  ModalContext :  1, 0,
|  Worker-4 :  0, 1,
|  Worker-3 :  -1, 0,
|  -------
```
The row following the "::" lists the locks. A lock is either an _ISchedulingRule_ or an explicit _ILock_. Each subsequent row lists the relationships between a thread and its acquired locks (&gt; 0), the locks it is waiting to acquire (-1), or the locks that it has no relationship (0). In the above, the workspace root ({@code R/}) has been acquired by <em>ModalContext</em> but <em>Worker-3</em> would like to acquire it, and <em>Worker-4</em> has acquired <em>ILock #4</em>.

Another example shows a single thread with the workspace root and lock 0.
```
|  :: 
|  R/, OrderedLock (0),
|  Worker-4 :  1, 1,
|  -------
```

### CODE REVIEW

**Comment by briandealwis** [comment]
Oh jeesh, another deadlock.  My suspicion here is that the Dependency Graph update job *had already been scheduled* before we turned off the updates.  So we need to change our pattern to:
```
    try {
      IDependencyGraph.INSTANCE.preUpdate();
      Job.getJobManager().join(DependencyGraphImpl.GRAPH_UPDATE_JOB_FAMILY, null);

      // do your thing
    } finally {
      IDependencyGraph.INSTANCE.postUpdate();
   }
```

Here is the relevant part of the deadlock:
```
+-------------------------------------------------------------------------------
| STACK DUMP @ 4.001 min: testAppEngineRuntimeAdded(com.google.cloud.tools.eclipse.appengine.newproject.CreateAppEngineStandardWtpProjectTest)
|
| "main" [1] TIMED_WAITING on org.eclipse.core.internal.jobs.Semaphore@1c2caaad
|     at java.lang.Object.wait(Native Method)
|     - waiting on org.eclipse.core.internal.jobs.Semaphore@1c2caaad
|     at org.eclipse.core.internal.jobs.Semaphore.acquire(Semaphore.java:39)
|     at org.eclipse.core.internal.jobs.OrderedLock.doAcquire(OrderedLock.java:170)
|     at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:106)
|     at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:82)
|     at org.eclipse.wst.common.componentcore.internal.StructureEdit.getComponentModelRoot(StructureEdit.java:467)
|     at org.eclipse.wst.common.componentcore.internal.StructureEdit.getWorkbenchModules(StructureEdit.java:506)
|     at org.eclipse.wst.common.componentcore.internal.StructureEdit.getComponent(StructureEdit.java:949)
|     at org.eclipse.wst.common.componentcore.internal.resources.VirtualComponent.createResource(VirtualComponent.java:124)
|     at org.eclipse.wst.common.componentcore.internal.resources.VirtualComponent.addManagedResource(VirtualComponent.java:100)
|     at org.eclipse.wst.common.componentcore.internal.resources.VirtualComponent.initializeResource(VirtualComponent.java:114)
|     at org.eclipse.wst.common.componentcore.internal.resources.VirtualComponent.<init>(VirtualComponent.java:148)
|     at org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager.createComponent(ComponentImplManager.java:250)
|     - locked org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager@4e13af1b
|     at org.eclipse.wst.common.componentcore.ComponentCore.createComponent(ComponentCore.java:84)
|     at org.eclipse.jst.j2ee.web.project.facet.WebFacetInstallDelegate.setContentPropertyIfNeeded(WebFacetInstallDelegate.java:221)
|     at org.eclipse.jst.j2ee.web.project.facet.WebFacetInstallDelegate.execute(WebFacetInstallDelegate.java:86)
|     at org.eclipse.wst.common.project.facet.core.internal.FacetedProject.callDelegate(FacetedProject.java:1477)
|     at org.eclipse.wst.common.project.facet.core.internal.FacetedProject.modifyInternal(FacetedProject.java:441)
|     at org.eclipse.wst.common.project.facet.core.internal.FacetedProject.mergeChangesInternal(FacetedProject.java:1181)
|     at org.eclipse.wst.common.project.facet.core.internal.FacetedProject.access$2(FacetedProject.java:1117)
|     at org.eclipse.wst.common.project.facet.core.internal.FacetedProject$1.run(FacetedProject.java:324)
|     at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2240)
|     at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2267)
|     at org.eclipse.wst.common.project.facet.core.internal.FacetedProject.modify(FacetedProject.java:339)
|     at com.google.cloud.tools.eclipse.appengine.facets.AppEngineStandardFacet.installAppEngineFacet(AppEngineStandardFacet.java:159)
|     at com.google.cloud.tools.eclipse.appengine.newproject.CreateAppEngineStandardWtpProject.execute(CreateAppEngineStandardWtpProject.java:95)
|     at com.google.cloud.tools.eclipse.appengine.newproject.CreateAppEngineStandardWtpProjectTest.testAppEngineRuntimeAdded(CreateAppEngineStandardWtpProjectTest.java:122)
|     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
|     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
|     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
|     at java.lang.reflect.Method.invoke(Method.java:483)
|     at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
|     at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
|     at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
|     at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
|     at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
|     at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
|     at com.google.cloud.tools.eclipse.test.util.ThreadDumpingWatchdog$1.evaluate(ThreadDumpingWatchdog.java:69)
|     at org.junit.rules.RunRules.evaluate(RunRules.java:20)
|     at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
|     at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
|     at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
|     at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
|     at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
|     at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
|     at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
|     at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
|     at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
|     at org.mockito.internal.runners.JUnit45AndHigherRunnerImpl.run(JUnit45AndHigherRunnerImpl.java:37)
|     at org.mockito.runners.MockitoJUnitRunner.run(MockitoJUnitRunner.java:62)
|     at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
|     at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
|     at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
|     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
|     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
|     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
|     at java.lang.reflect.Method.invoke(Method.java:483)
|     at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray2(ReflectionUtils.java:208)
|     at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:156)
|     at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:82)
|     at org.eclipse.tycho.surefire.osgibooter.OsgiSurefireBooter.run(OsgiSurefireBooter.java:91)
|     at org.eclipse.tycho.surefire.osgibooter.AbstractUITestApplication$1.run(AbstractUITestApplication.java:35)
|     at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:37)
|     at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:182)
|     - locked org.eclipse.swt.widgets.RunnableLock@1b33aa3c
|     at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4468)
|     at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4081)
|     at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1133)
|     at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
|     at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1022)
|     at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
|     at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:684)
|     at org.eclipse.ui.internal.Workbench$$Lambda$18/1684541910.run(Unknown Source)
|     at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
|     at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:598)
|     at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
|     at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:138)
|     at org.eclipse.tycho.surefire.osgibooter.UITestApplication.runApplication(UITestApplication.java:31)
|     at org.eclipse.tycho.surefire.osgibooter.AbstractUITestApplication.run(AbstractUITestApplication.java:120)
|     at org.eclipse.tycho.surefire.osgibooter.UITestApplication.start(UITestApplication.java:37)
|     at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
|     at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
|     at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
|     at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
|     at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
|     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
|     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
|     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
|     at java.lang.reflect.Method.invoke(Method.java:483)
|     at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:653)
|     at org.eclipse.equinox.launcher.Main.basicRun(Main.java:590)
|     at org.eclipse.equinox.launcher.Main.run(Main.java:1499)
|     at org.eclipse.equinox.launcher.Main.main(Main.java:1472)
| 
| "Worker-6" [90] BLOCKED on org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager@4e13af1b owned by 'main [id:1]
|     at org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager.createComponent(ComponentImplManager.java:223)
|     - blocked on org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager@4e13af1b
|     at org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager.createComponent(ComponentImplManager.java:212)
|     at org.eclipse.wst.common.componentcore.ComponentCore.createComponent(ComponentCore.java:64)
|     at org.eclipse.wst.common.componentcore.internal.builder.DependencyGraphImpl$GraphUpdateJob$1.run(DependencyGraphImpl.java:494)
|     at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
|     at org.eclipse.wst.common.componentcore.internal.builder.DependencyGraphImpl$GraphUpdateJob.run(DependencyGraphImpl.java:462)
|     at org.eclipse.core.internal.jobs.Worker.run(Worker.java:56)
```

File: plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ThreadDumpingWatchdog.java [hunk 1]
```diff
@@ -112,14 +121,62 @@ public int compare(ThreadInfo o1, ThreadInfo o2) {
...
+        // don't output anything if no locks are held
+        return;
+      }
+      // locks is an instanceof DeadlockDetector
+      Object locks = ReflectionUtil.getField(manager, "locks", Object.class);
+      String debugOutput = ReflectionUtil.invoke(locks, "toDebugString", String.class);
+      sb.append("\n").append(linePrefix);
+      sb.append("\n").append(linePrefix).append("Eclipse Locks:");
+      sb.append("\n").append(linePrefix).append(debugOutput.replace("\n", "\n" + linePrefix));
+    } catch (Exception e) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ThreadDumpingWatchdog.java]: could this be RuntimeException instead? or more specific exceptions?

File: plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ThreadDumpingWatchdog.java [hunk 2]
```diff
@@ -112,14 +121,62 @@ public int compare(ThreadInfo o1, ThreadInfo o2) {
...
+   * Eclipse Locks:
+   *  :: 
+   *  R/, OrderedLock (4),
+   *  ModalContext :  1, 0,
+   *  Worker-4 :  0, 1,
+   *  Worker-3 :  -1, 0,
+   *  -------
+   * </pre>
+   * 
+   * The first row following "::" list the locks. A lock is either an
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ThreadDumpingWatchdog.java]: "lists the locks".

**Comment by briandealwis** [comment]
> Might be worth filing a request at Eclipse to expose more of this.

Agreed.  I'd love if there could be a way so that the locks and rules could be exposed so they're seen in the `jstack`-like output.

**Comment by briandealwis** [comment]
Build failing because of #1363, #1373.  But they demonstrate that this change works!

**Pull request merged**